{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "expected-boutique",
   "metadata": {},
   "source": [
    "# Introduction\n",
    "\n",
    "This is a simplification of my figuring out how to submit seqspec files to the IGVF DACC.\n",
    "\n",
    "- [Python environment](#Setup)\n",
    "- [Seqspec template function](#Template)\n",
    "- [Working out steps needed to create seqspec objects](#Working-out-steps-needed-to-create-seqspec-objects)\n",
    "- [Exploring boto3](#Exploring-boto3)\n",
    "- [Seqspec submission functions](#Seqspec-Submission-functions)\n",
    "- [Create seqspec objects for remaining fastqs](#Create-seqspec-objects-for-remaining-fastqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6789786",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "26d0ec49",
   "metadata": {},
   "source": [
    "Information about barcodes.\n",
    "\n",
    "cell barcode whitelists for 10x multiome GEX and ATAC:\n",
    "- atac_737K-arc-v1.txt.gz\n",
    "- gex_737K-arc-v1.txt.gz\n",
    "\n",
    "MULTI-seq barcodes (custom)\n",
    "- MULTIbc_onlist.txt \n",
    "\n",
    "MULTI-seq I7 indices (custom):\n",
    "- index7_onlist_MULTI.txt \n",
    "\n",
    "10x Multiome \n",
    "- snRNA-seq I5 and I7 indices: https://cdn.10xgenomics.com/raw/upload/v1655151897/support/in-line%20documents/Dual_Index_Kit_TT_Set_A.csv\n",
    "  - i7_dual_index_set_a.txt (10 bp)\n",
    "  - i5_dual_index_set_a.txt (10 bp)\n",
    "- ATAC-seq I7 indices: https://cdn.10xgenomics.com/raw/upload/v1655155348/support/in-line%20documents/Single_Index_Kit_N_Set_A.csv\n",
    "  - i7_single_index_set_a.txt (8 bp)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "systematic-binary",
   "metadata": {},
   "source": [
    "# Setup\n",
    "\n",
    "First we start with general imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "irish-queue",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import hashlib\n",
    "from io import StringIO, BytesIO\n",
    "import logging\n",
    "import numpy\n",
    "import os\n",
    "import pandas\n",
    "from pathlib import Path\n",
    "import requests\n",
    "import sys\n",
    "import tarfile\n",
    "import tqdm\n",
    "from urllib.parse import urlparse\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "least-senate",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    import boto3\n",
    "except ImportError:\n",
    "    !{sys.executable} -m pip install --user boto3\n",
    "    import boto3\n",
    "    \n",
    "from botocore.exceptions import ClientError\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "double-break",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    from jinja2 import Environment\n",
    "except ImportError:\n",
    "    !{sys.executable} -m pip install --user jinja2\n",
    "    from jinja2 import Environment\n",
    "\n",
    "from jinja2 import FileSystemLoader, select_autoescape, Undefined, StrictUndefined, make_logging_undefined\n",
    "\n",
    "logger = logging.getLogger(__name__)\n",
    "LoggingUndefined = make_logging_undefined(\n",
    "    logger=logger,\n",
    "    base=Undefined\n",
    ")\n",
    "\n",
    "env = Environment(\n",
    "    loader=FileSystemLoader(\"templates\"),\n",
    "    autoescape=select_autoescape(),\n",
    "    undefined=LoggingUndefined,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bibliographic-jonathan",
   "metadata": {},
   "source": [
    "I want to be able to use the seqspec validator while I was writing my seqspec file.\n",
    "\n",
    "I have the repository checked out into ~/proj/seqspec. This block should either import it for me, or install it if someone elese runs it."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "usual-census",
   "metadata": {},
   "source": [
    "Import pieces of seqspec that we need for this notebook."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fleet-pillow",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    import seqspec\n",
    "except ImportError:\n",
    "    seqspec_root = Path(\"~/proj/seqspec\").expanduser()\n",
    "    if seqspec_root.exists() and str(seqspec_root) not in sys.path:\n",
    "        sys.path.append(str(seqspec_root))\n",
    "    else:\n",
    "        !{sys.executable} -m pip install --user seqspec\n",
    "    import seqspec\n",
    "    \n",
    "from seqspec.Assay import Assay\n",
    "from seqspec.Region import Region\n",
    "from seqspec.Region import Onlist\n",
    "from seqspec.utils import load_spec_stream\n",
    "\n",
    "#from seqspec.seqspec_index import run_index, get_index\n",
    "#from seqspec.seqspec_print import run_print_sequence_spec, run_print_library_tree, run_print_library_png\n",
    "#from seqspec.seqspec_onlist import run_list_onlists, run_onlist_read, run_find_by_type    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4bd5a7fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "from seqspecgen.igvf import (\n",
    "    get_sequence_file_info,\n",
    "    get_barcode_info,\n",
    ")\n",
    "from seqspecgen.util import seqspec_validate"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "sharp-quick",
   "metadata": {},
   "source": [
    "I have my own API for interacting with the IGVF database server (which is very much like the old ENCODE database server)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "vocal-nutrition",
   "metadata": {},
   "outputs": [],
   "source": [
    "try:\n",
    "    from encoded_client import encoded\n",
    "except ImportError:\n",
    "    encoded_root = Path(\"~/proj/encoded_client\").expanduser()\n",
    "    if encoded_root.exists() and str(encoded_root) not in sys.path:\n",
    "        sys.path.append(str(encoded_root))\n",
    "    else:\n",
    "        !{sys.executable} -m pip install --user encoded_client\n",
    "        \n",
    "    from encoded_client import encoded"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "increasing-entity",
   "metadata": {},
   "source": [
    "encoded_client will pull submitter credentials from either DCC_API_KEY and DCC_SECRET_KEY or from a .netrc file loaded from your home directory. (replacing the {DCC_API_KEY} and {DCC_SECRET_KEY} strings with your specific values.)\n",
    "\n",
    "The format of a .netrc file is a plain text file with records of the format:\n",
    "\n",
    "<pre>machine api.sandbox.igvf.org login {DCC_API_KEY} password {DCC_SECRET_KEY}</pre>\n",
    "\n",
    "Or api.data.igvf.org\n",
    "\n",
    "(it's also possible to list the fields on separate lines, but I think it's easier to read when they're on one line)\n",
    "\n",
    "or after creating the server object call:\n",
    "\n",
    "<pre>server.username = \"{DCC_API_KEY}\"\n",
    "server.password = \"{DCC_SECRET_KEY}\"</pre>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "valuable-meditation",
   "metadata": {},
   "source": [
    "## submission variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "alleged-hebrew",
   "metadata": {},
   "outputs": [],
   "source": [
    "#server_name = \"api.sandbox.igvf.org\"\n",
    "server_name = \"api.data.igvf.org\"\n",
    "award = \"/awards/HG012076/\"\n",
    "lab = \"/labs/ansuman-satpathy/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "usual-greenhouse",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "human-coordinate",
   "metadata": {},
   "outputs": [],
   "source": [
    "server = encoded.ENCODED(server_name)\n",
    "igvf_validator = encoded.DCCValidator(server)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf74e305",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "regional-remedy",
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_spec(filename):\n",
    "    with open(filename, \"rt\") as instream:\n",
    "        data = yaml.load(instream, Loader=yaml.Loader)\n",
    "        for r in data.assay_spec:\n",
    "            r.set_parent_id(None)\n",
    "    return data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "allied-cargo",
   "metadata": {},
   "source": [
    "# Template"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "processed-values",
   "metadata": {},
   "source": [
    "First build up lists of barcodes onlists needed for this protocol the names will be passed to the template."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cf15d013",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>identifier</th>\n",
       "      <th>modality</th>\n",
       "      <th>sequence_kit</th>\n",
       "      <th>sequence_protocol</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>R3</th>\n",
       "      <th>I1</th>\n",
       "      <th>I2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ansuman-satpathy:igvf_exp0_atac_10x1_NGS1</td>\n",
       "      <td>atac</td>\n",
       "      <td>NovaSeq 6000 S4 Reagent Kit v1.5</td>\n",
       "      <td>Illumina NovaSeq 6000 (EFO:0008637)</td>\n",
       "      <td>IGVFFI3568SBTV</td>\n",
       "      <td>IGVFFI1986FAVM</td>\n",
       "      <td>IGVFFI8652MKWJ</td>\n",
       "      <td>IGVFFI5795ORFY</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ansuman-satpathy:igvf_exp0_multi_10x1_NGS1</td>\n",
       "      <td>multi</td>\n",
       "      <td>NovaSeq 6000 S4 Reagent Kit v1.5</td>\n",
       "      <td>Illumina NovaSeq 6000 (EFO:0008637)</td>\n",
       "      <td>IGVFFI5411OJRK</td>\n",
       "      <td>IGVFFI4642YKWI</td>\n",
       "      <td>IGVFFI5223HYRP</td>\n",
       "      <td>IGVFFI6088QALC</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ansuman-satpathy:igvf_exp0_rna_10x1_NGS1</td>\n",
       "      <td>rna</td>\n",
       "      <td>NovaSeq 6000 S4 Reagent Kit v1.5</td>\n",
       "      <td>Illumina NovaSeq 6000 (EFO:0008637)</td>\n",
       "      <td>IGVFFI5758URSR</td>\n",
       "      <td>IGVFFI1215RCAP</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVFFI4517STPG</td>\n",
       "      <td>IGVFFI8677MRVW</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ansuman-satpathy:igvf_exp1_atac_10x1_NGS1</td>\n",
       "      <td>atac</td>\n",
       "      <td>NovaSeq X Series 10B Reagent Kit</td>\n",
       "      <td>Illumina NovaSeq X (NTR:0000765)</td>\n",
       "      <td>IGVFFI1415ZCCR</td>\n",
       "      <td>IGVFFI9193AKLW</td>\n",
       "      <td>IGVFFI4166FQWH</td>\n",
       "      <td>IGVFFI9476DBSV</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ansuman-satpathy:igvf_exp1_atac_10x1_NGS2</td>\n",
       "      <td>atac</td>\n",
       "      <td>NovaSeq X Series 10B Reagent Kit</td>\n",
       "      <td>Illumina NovaSeq X (NTR:0000765)</td>\n",
       "      <td>IGVFFI2636JAAE</td>\n",
       "      <td>IGVFFI0014KSVS</td>\n",
       "      <td>IGVFFI9875PICN</td>\n",
       "      <td>IGVFFI6845HIAS</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>408</th>\n",
       "      <td>ansuman-satpathy:igvf_mouse_bridge_114_rna</td>\n",
       "      <td>rna</td>\n",
       "      <td>NovaSeq 6000 S4 Reagent Kit v1.5</td>\n",
       "      <td>Illumina NovaSeq 6000 (EFO:0008637)</td>\n",
       "      <td>IGVFFI3400EIAB</td>\n",
       "      <td>IGVFFI5674QKJY</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVFFI1603WDWP</td>\n",
       "      <td>IGVFFI1522TXOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>ansuman-satpathy:igvf_mouse_bridge_115_atac</td>\n",
       "      <td>atac</td>\n",
       "      <td>NovaSeq 6000 S4 Reagent Kit v1.5</td>\n",
       "      <td>Illumina NovaSeq 6000 (EFO:0008637)</td>\n",
       "      <td>IGVFFI5559EVYB</td>\n",
       "      <td>IGVFFI5706AEPS</td>\n",
       "      <td>IGVFFI1771YVVX</td>\n",
       "      <td>IGVFFI1992CETB</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>410</th>\n",
       "      <td>ansuman-satpathy:igvf_mouse_bridge_115_rna</td>\n",
       "      <td>rna</td>\n",
       "      <td>NovaSeq 6000 S4 Reagent Kit v1.5</td>\n",
       "      <td>Illumina NovaSeq 6000 (EFO:0008637)</td>\n",
       "      <td>IGVFFI7761PXYZ</td>\n",
       "      <td>IGVFFI2504VZJX</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVFFI0849BAAC</td>\n",
       "      <td>IGVFFI3865CUKI</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>411</th>\n",
       "      <td>ansuman-satpathy:igvf_mouse_bridge_116_atac</td>\n",
       "      <td>atac</td>\n",
       "      <td>NovaSeq 6000 S4 Reagent Kit v1.5</td>\n",
       "      <td>Illumina NovaSeq 6000 (EFO:0008637)</td>\n",
       "      <td>IGVFFI7842VQRM</td>\n",
       "      <td>IGVFFI0599BUEC</td>\n",
       "      <td>IGVFFI0780GMLK</td>\n",
       "      <td>IGVFFI9383DDZQ</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>412</th>\n",
       "      <td>ansuman-satpathy:igvf_mouse_bridge_116_rna</td>\n",
       "      <td>rna</td>\n",
       "      <td>NovaSeq 6000 S4 Reagent Kit v1.5</td>\n",
       "      <td>Illumina NovaSeq 6000 (EFO:0008637)</td>\n",
       "      <td>IGVFFI0666JVDJ</td>\n",
       "      <td>IGVFFI5154UDRD</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IGVFFI8523UBIS</td>\n",
       "      <td>IGVFFI0365KTIZ</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>413 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      identifier modality  \\\n",
       "0      ansuman-satpathy:igvf_exp0_atac_10x1_NGS1     atac   \n",
       "1     ansuman-satpathy:igvf_exp0_multi_10x1_NGS1    multi   \n",
       "2       ansuman-satpathy:igvf_exp0_rna_10x1_NGS1      rna   \n",
       "3      ansuman-satpathy:igvf_exp1_atac_10x1_NGS1     atac   \n",
       "4      ansuman-satpathy:igvf_exp1_atac_10x1_NGS2     atac   \n",
       "..                                           ...      ...   \n",
       "408   ansuman-satpathy:igvf_mouse_bridge_114_rna      rna   \n",
       "409  ansuman-satpathy:igvf_mouse_bridge_115_atac     atac   \n",
       "410   ansuman-satpathy:igvf_mouse_bridge_115_rna      rna   \n",
       "411  ansuman-satpathy:igvf_mouse_bridge_116_atac     atac   \n",
       "412   ansuman-satpathy:igvf_mouse_bridge_116_rna      rna   \n",
       "\n",
       "                         sequence_kit                    sequence_protocol  \\\n",
       "0    NovaSeq 6000 S4 Reagent Kit v1.5  Illumina NovaSeq 6000 (EFO:0008637)   \n",
       "1    NovaSeq 6000 S4 Reagent Kit v1.5  Illumina NovaSeq 6000 (EFO:0008637)   \n",
       "2    NovaSeq 6000 S4 Reagent Kit v1.5  Illumina NovaSeq 6000 (EFO:0008637)   \n",
       "3    NovaSeq X Series 10B Reagent Kit     Illumina NovaSeq X (NTR:0000765)   \n",
       "4    NovaSeq X Series 10B Reagent Kit     Illumina NovaSeq X (NTR:0000765)   \n",
       "..                                ...                                  ...   \n",
       "408  NovaSeq 6000 S4 Reagent Kit v1.5  Illumina NovaSeq 6000 (EFO:0008637)   \n",
       "409  NovaSeq 6000 S4 Reagent Kit v1.5  Illumina NovaSeq 6000 (EFO:0008637)   \n",
       "410  NovaSeq 6000 S4 Reagent Kit v1.5  Illumina NovaSeq 6000 (EFO:0008637)   \n",
       "411  NovaSeq 6000 S4 Reagent Kit v1.5  Illumina NovaSeq 6000 (EFO:0008637)   \n",
       "412  NovaSeq 6000 S4 Reagent Kit v1.5  Illumina NovaSeq 6000 (EFO:0008637)   \n",
       "\n",
       "                 R1              R2              R3              I1  \\\n",
       "0    IGVFFI3568SBTV  IGVFFI1986FAVM  IGVFFI8652MKWJ  IGVFFI5795ORFY   \n",
       "1    IGVFFI5411OJRK  IGVFFI4642YKWI  IGVFFI5223HYRP  IGVFFI6088QALC   \n",
       "2    IGVFFI5758URSR  IGVFFI1215RCAP             NaN  IGVFFI4517STPG   \n",
       "3    IGVFFI1415ZCCR  IGVFFI9193AKLW  IGVFFI4166FQWH  IGVFFI9476DBSV   \n",
       "4    IGVFFI2636JAAE  IGVFFI0014KSVS  IGVFFI9875PICN  IGVFFI6845HIAS   \n",
       "..              ...             ...             ...             ...   \n",
       "408  IGVFFI3400EIAB  IGVFFI5674QKJY             NaN  IGVFFI1603WDWP   \n",
       "409  IGVFFI5559EVYB  IGVFFI5706AEPS  IGVFFI1771YVVX  IGVFFI1992CETB   \n",
       "410  IGVFFI7761PXYZ  IGVFFI2504VZJX             NaN  IGVFFI0849BAAC   \n",
       "411  IGVFFI7842VQRM  IGVFFI0599BUEC  IGVFFI0780GMLK  IGVFFI9383DDZQ   \n",
       "412  IGVFFI0666JVDJ  IGVFFI5154UDRD             NaN  IGVFFI8523UBIS   \n",
       "\n",
       "                 I2  \n",
       "0               NaN  \n",
       "1               NaN  \n",
       "2    IGVFFI8677MRVW  \n",
       "3               NaN  \n",
       "4               NaN  \n",
       "..              ...  \n",
       "408  IGVFFI1522TXOM  \n",
       "409             NaN  \n",
       "410  IGVFFI3865CUKI  \n",
       "411             NaN  \n",
       "412  IGVFFI0365KTIZ  \n",
       "\n",
       "[413 rows x 9 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata = pandas.read_csv(\"satpathy_seqspec_files.tsv\", sep=\"\\t\")\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "807f599b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "pointed-worcester",
   "metadata": {},
   "outputs": [],
   "source": [
    "barcode_sets = {\n",
    "    \"rna\": {},\n",
    "    \"atac\": {},\n",
    "    \"multi\": {},\n",
    "}\n",
    "\n",
    "barcode_sets[\"rna\"].update(get_barcode_info(server, \"cell_barcode\", \"IGVFFI8751YQRY\"))\n",
    "barcode_sets[\"rna\"].update(get_barcode_info(server, \"i7_rnaseq_index\", \"IGVFFI4565KANH\"))\n",
    "barcode_sets[\"rna\"].update(get_barcode_info(server, \"i5_rnaseq_index\", \"IGVFFI4565KANH\"))\n",
    "\n",
    "barcode_sets[\"atac\"].update(get_barcode_info(server, \"cell_barcode\", \"IGVFFI7587TJLC\"))\n",
    "barcode_sets[\"atac\"].update(get_barcode_info(server, \"i7_atac_index\", \"IGVFFI1608YDWY\"))\n",
    "\n",
    "barcode_sets[\"multi\"].update(get_barcode_info(server, \"cell_barcode\", \"IGVFFI8751YQRY\"))\n",
    "barcode_sets[\"multi\"].update(get_barcode_info(server, \"multi_index\", \"IGVFFI5410CKWT\"))\n",
    "barcode_sets[\"multi\"].update(get_barcode_info(server, \"i7_multi_index\", \"IGVFFI3231EAMC\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b4a67076",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'rna': {'cell_barcode_file_id': 'IGVFFI8751YQRY',\n",
       "  'cell_barcode_file_name': 'IGVFFI8751YQRY.tsv.gz',\n",
       "  'cell_barcode_file_size': 2142553,\n",
       "  'cell_barcode_url': 'https://api.data.igvf.org/tabular-files/IGVFFI8751YQRY/@@download/IGVFFI8751YQRY.tsv.gz',\n",
       "  'cell_barcode_md5': '95ca0127739965ae57d09c6c1cfb20b4',\n",
       "  'i7_rnaseq_index_file_id': 'IGVFFI4565KANH',\n",
       "  'i7_rnaseq_index_file_name': 'IGVFFI4565KANH.csv.gz',\n",
       "  'i7_rnaseq_index_file_size': 1752,\n",
       "  'i7_rnaseq_index_url': 'https://api.data.igvf.org/tabular-files/IGVFFI4565KANH/@@download/IGVFFI4565KANH.csv.gz',\n",
       "  'i7_rnaseq_index_md5': '9bbbf8c40bf751cea0505fc85a66ab08',\n",
       "  'i5_rnaseq_index_file_id': 'IGVFFI4565KANH',\n",
       "  'i5_rnaseq_index_file_name': 'IGVFFI4565KANH.csv.gz',\n",
       "  'i5_rnaseq_index_file_size': 1752,\n",
       "  'i5_rnaseq_index_url': 'https://api.data.igvf.org/tabular-files/IGVFFI4565KANH/@@download/IGVFFI4565KANH.csv.gz',\n",
       "  'i5_rnaseq_index_md5': '9bbbf8c40bf751cea0505fc85a66ab08'},\n",
       " 'atac': {'cell_barcode_file_id': 'IGVFFI7587TJLC',\n",
       "  'cell_barcode_file_name': 'IGVFFI7587TJLC.tsv.gz',\n",
       "  'cell_barcode_file_size': 2465078,\n",
       "  'cell_barcode_url': 'https://api.data.igvf.org/tabular-files/IGVFFI7587TJLC/@@download/IGVFFI7587TJLC.tsv.gz',\n",
       "  'cell_barcode_md5': '91f5bd173373fa1815830444480236fb',\n",
       "  'i7_atac_index_file_id': 'IGVFFI1608YDWY',\n",
       "  'i7_atac_index_file_name': 'IGVFFI1608YDWY.csv.gz',\n",
       "  'i7_atac_index_file_size': 1658,\n",
       "  'i7_atac_index_url': 'https://api.data.igvf.org/tabular-files/IGVFFI1608YDWY/@@download/IGVFFI1608YDWY.csv.gz',\n",
       "  'i7_atac_index_md5': 'db54507732297fafea74bacfcc203238'},\n",
       " 'multi': {'cell_barcode_file_id': 'IGVFFI8751YQRY',\n",
       "  'cell_barcode_file_name': 'IGVFFI8751YQRY.tsv.gz',\n",
       "  'cell_barcode_file_size': 2142553,\n",
       "  'cell_barcode_url': 'https://api.data.igvf.org/tabular-files/IGVFFI8751YQRY/@@download/IGVFFI8751YQRY.tsv.gz',\n",
       "  'cell_barcode_md5': '95ca0127739965ae57d09c6c1cfb20b4',\n",
       "  'multi_index_file_id': 'IGVFFI5410CKWT',\n",
       "  'multi_index_file_name': 'IGVFFI5410CKWT.tsv.gz',\n",
       "  'multi_index_file_size': 664,\n",
       "  'multi_index_url': 'https://api.data.igvf.org/tabular-files/IGVFFI5410CKWT/@@download/IGVFFI5410CKWT.tsv.gz',\n",
       "  'multi_index_md5': '78f0ed53d9dc313a09a6dc7e46ccc2ce',\n",
       "  'i7_multi_index_file_id': 'IGVFFI3231EAMC',\n",
       "  'i7_multi_index_file_name': 'IGVFFI3231EAMC.tsv.gz',\n",
       "  'i7_multi_index_file_size': 124,\n",
       "  'i7_multi_index_url': 'https://api.data.igvf.org/tabular-files/IGVFFI3231EAMC/@@download/IGVFFI3231EAMC.tsv.gz',\n",
       "  'i7_multi_index_md5': '91bb39e4be46f1ffb51caa4eb54f8897'}}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "barcode_sets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "handled-nickname",
   "metadata": {},
   "source": [
    "Define the the file information needed to define this seqspec."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "behavioral-dressing",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# load the template and instantiate it with the context for this specific measurement set.\n",
    "# this version needs the libspec branch that separates the sequence and library specifications\n",
    "\n",
    "def create_seqspec(row, verbose=False):\n",
    "    #template_name = \"parse-wt-mega-v2-single-index-libspec-1.yaml.j2\"\n",
    "    modality = row[\"modality\"]\n",
    "    template_name = \"igvf-multiseq-{modality}.yaml.j2\".format(modality=modality)\n",
    "    template = env.get_template(template_name)\n",
    "\n",
    "    # Build context\n",
    "    context = {}\n",
    "    # Merge in the barcode information\n",
    "    context.update(barcode_sets[modality])\n",
    "\n",
    "    context.update(row.to_dict())\n",
    "\n",
    "    for column in [\"R1\", \"R2\", \"R3\", \"I1\", \"I2\"]:\n",
    "        if pandas.notnull(row[column]):\n",
    "            accession = row[column]\n",
    "            context.update(get_sequence_file_info(server, accession, column))\n",
    "            #lengths = get_read_lengths(server, row[column], reads_to_check=10)\n",
    "            #context[f\"{column}_min_len\"] = min(lengths)\n",
    "            #context[f\"{column}_max_len\"] = max(lengths)\n",
    "\n",
    "    example_yaml = template.render(context)\n",
    "\n",
    "    # validate the generated seqspec file.\n",
    "    example_spec = load_spec_stream(StringIO(example_yaml))\n",
    "    seqspec_validate(example_spec.to_dict())\n",
    "\n",
    "    example_spec.update_spec()\n",
    "    if verbose:\n",
    "        print(example_spec.to_YAML())\n",
    "    return example_spec"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "832534e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "identifier           ansuman-satpathy:igvf_exp0_multi_10x1_NGS1\n",
       "modality                                                  multi\n",
       "sequence_kit                   NovaSeq 6000 S4 Reagent Kit v1.5\n",
       "sequence_protocol           Illumina NovaSeq 6000 (EFO:0008637)\n",
       "R1                                               IGVFFI5411OJRK\n",
       "R2                                               IGVFFI4642YKWI\n",
       "R3                                               IGVFFI5223HYRP\n",
       "I1                                               IGVFFI6088QALC\n",
       "I2                                                          NaN\n",
       "Name: 1, dtype: object"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata[metadata[\"modality\"] == \"multi\"].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cognitive-admission",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "# rna example\n",
    "#example_spec = create_seqspec(metadata[metadata[\"modality\"] == \"rna\"].iloc[0])\n",
    "\n",
    "# atac example\n",
    "#example_spec = create_seqspec(metadata[metadata[\"modality\"] == \"atac\"].iloc[0])\n",
    "\n",
    "# multi example\n",
    "example_spec = create_seqspec(metadata[metadata[\"modality\"] == \"multi\"].iloc[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "electrical-immune",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "!Assay\n",
      "seqspec_version: 0.3.0\n",
      "assay_id: 10x-ATAC-RNA-MULTI\n",
      "name: 10x-ATAC-RNA-MULTI/Illumina\n",
      "doi: https://doi.org/10.1038/s41592-019-0433-8\n",
      "date: 17 June 2019\n",
      "description: ansuman-satpathy:igvf_exp0_multi_10x1_NGS1 MULTI-seq\n",
      "modalities:\n",
      "- tag\n",
      "lib_struct: https://igvf.github.io/seqspec/\n",
      "library_protocol: Custom\n",
      "library_kit: Illumina Truseq Dual Index\n",
      "sequence_protocol: Illumina NovaSeq 6000 (EFO:0008637)\n",
      "sequence_kit: NovaSeq 6000 S4 Reagent Kit v1.5\n",
      "sequence_spec:\n",
      "- !Read\n",
      "  read_id: IGVFFI5411OJRK\n",
      "  name: Read 1\n",
      "  modality: tag\n",
      "  primer_id: multi-truseq_read1\n",
      "  min_len: 50\n",
      "  max_len: 50\n",
      "  strand: pos\n",
      "  files:\n",
      "  - !File\n",
      "    file_id: IGVFFI5411OJRK\n",
      "    filename: IGVFFI5411OJRK.fastq.gz\n",
      "    filetype: fastq\n",
      "    filesize: 3744591369\n",
      "    url: https://api.data.igvf.org/sequence-files/IGVFFI5411OJRK/@@download/IGVFFI5411OJRK.fastq.gz\n",
      "    urltype: https\n",
      "    md5: 1e97130b5dfa0cdd859e95b39af60494\n",
      "- !Read\n",
      "  read_id: IGVFFI6088QALC\n",
      "  name: Index 1 (i7 index)\n",
      "  modality: tag\n",
      "  primer_id: multi-truseq_read2\n",
      "  min_len: 8\n",
      "  max_len: 8\n",
      "  strand: pos\n",
      "  files:\n",
      "  - !File\n",
      "    file_id: IGVFFI6088QALC\n",
      "    filename: IGVFFI6088QALC.fastq.gz\n",
      "    filetype: IGVFFI6088QALC.fastq.gz\n",
      "    filesize: 1090207721\n",
      "    url: https://api.data.igvf.org/sequence-files/IGVFFI6088QALC/@@download/IGVFFI6088QALC.fastq.gz\n",
      "    urltype: https\n",
      "    md5: b9bbf6f4133d80e5062a2bed6cf37bb2\n",
      "- !Read\n",
      "  read_id: IGVFFI4642YKWI\n",
      "  name: Read 2\n",
      "  modality: tag\n",
      "  primer_id: multi-truseq_read2\n",
      "  min_len: 24\n",
      "  max_len: 24\n",
      "  strand: neg\n",
      "  files:\n",
      "  - !File\n",
      "    file_id: IGVFFI4642YKWI\n",
      "    filename: IGVFFI4642YKWI.fastq.gz\n",
      "    filetype: fastq\n",
      "    filesize: 1839821154\n",
      "    url: https://api.data.igvf.org/sequence-files/IGVFFI4642YKWI/@@download/IGVFFI4642YKWI.fastq.gz\n",
      "    urltype: https\n",
      "    md5: e0cd5d86193f65d2dc20bab60adcf9d1\n",
      "library_spec:\n",
      "- !Region\n",
      "  parent_id: null\n",
      "  region_id: tag\n",
      "  region_type: tag\n",
      "  name: MULTI sample barcode\n",
      "  sequence_type: joined\n",
      "  sequence: AATGATACGGCGACCACCGAGATCTACACTCTTTCCCTACACGACGCTCTTCCGATCTNNNNNNNNNNNNNNNNXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXNNNNNNNNTGGAATTCTCGGGTGCCAAGGAACTCCAGTCACNNNNNNNNATCTCGTATGCCGTCTTCTGCTTG\n",
      "  min_len: 259\n",
      "  max_len: 259\n",
      "  onlist: null\n",
      "  regions:\n",
      "  - !Region\n",
      "    region_id: MULTI-illumina_p5\n",
      "    region_type: illumina_p5\n",
      "    name: Illumina P5\n",
      "    sequence_type: fixed\n",
      "    sequence: AATGATACGGCGACCACCGAGATCTACAC\n",
      "    min_len: 29\n",
      "    max_len: 29\n",
      "    onlist: null\n",
      "    regions: null\n",
      "    parent_id: tag\n",
      "  - !Region\n",
      "    region_id: multi-truseq_read1\n",
      "    region_type: truseq_read1\n",
      "    name: Truseq Read 1\n",
      "    sequence_type: fixed\n",
      "    sequence: TCTTTCCCTACACGACGCTCTTCCGATCT\n",
      "    min_len: 29\n",
      "    max_len: 29\n",
      "    onlist: null\n",
      "    regions: null\n",
      "    parent_id: tag\n",
      "  - !Region\n",
      "    region_id: barcode\n",
      "    region_type: barcode\n",
      "    name: Cell Barcode\n",
      "    sequence_type: onlist\n",
      "    sequence: NNNNNNNNNNNNNNNN\n",
      "    min_len: 16\n",
      "    max_len: 16\n",
      "    onlist: !Onlist\n",
      "      file_id: IGVFFI8751YQRY\n",
      "      filename: IGVFFI8751YQRY.tsv.gz\n",
      "      filetype: tsv\n",
      "      filesize: 2142553\n",
      "      url: https://api.data.igvf.org/tabular-files/IGVFFI8751YQRY/@@download/IGVFFI8751YQRY.tsv.gz\n",
      "      urltype: https\n",
      "      md5: 95ca0127739965ae57d09c6c1cfb20b4\n",
      "    regions: null\n",
      "    parent_id: tag\n",
      "  - !Region\n",
      "    region_id: umi\n",
      "    region_type: umi\n",
      "    name: UMI\n",
      "    sequence_type: random\n",
      "    sequence: XXXXXXXXXXXX\n",
      "    min_len: 12\n",
      "    max_len: 12\n",
      "    onlist: null\n",
      "    regions: null\n",
      "    parent_id: tag\n",
      "  - !Region\n",
      "    region_id: random_spacer\n",
      "    region_type: linker\n",
      "    name: MULTI-seq linker\n",
      "    sequence_type: random\n",
      "    sequence: XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX\n",
      "    min_len: 100\n",
      "    max_len: 100\n",
      "    onlist: null\n",
      "    regions: null\n",
      "    parent_id: tag\n",
      "  - !Region\n",
      "    region_id: MULTI-seq barcode\n",
      "    region_type: barcode\n",
      "    name: MULTI-seq barcode\n",
      "    sequence_type: onlist\n",
      "    sequence: NNNNNNNN\n",
      "    min_len: 8\n",
      "    max_len: 8\n",
      "    onlist: !Onlist\n",
      "      file_id: IGVFFI5410CKWT\n",
      "      filename: IGVFFI5410CKWT.tsv.gz\n",
      "      filetype: tsv\n",
      "      filesize: 664\n",
      "      url: https://api.data.igvf.org/tabular-files/IGVFFI5410CKWT/@@download/IGVFFI5410CKWT.tsv.gz\n",
      "      urltype: https\n",
      "      md5: 78f0ed53d9dc313a09a6dc7e46ccc2ce\n",
      "    regions: null\n",
      "    parent_id: tag\n",
      "  - !Region\n",
      "    region_id: multi-truseq_read2\n",
      "    region_type: truseq_read2\n",
      "    name: Truseq Small RNA Read 2\n",
      "    sequence_type: fixed\n",
      "    sequence: TGGAATTCTCGGGTGCCAAGGAACTCCAGTCAC\n",
      "    min_len: 33\n",
      "    max_len: 33\n",
      "    onlist: null\n",
      "    regions: null\n",
      "    parent_id: tag\n",
      "  - !Region\n",
      "    region_id: MULTI-index7\n",
      "    region_type: index7\n",
      "    name: index7\n",
      "    sequence_type: onlist\n",
      "    sequence: NNNNNNNN\n",
      "    min_len: 8\n",
      "    max_len: 8\n",
      "    onlist: !Onlist\n",
      "      file_id: IGVFFI3231EAMC\n",
      "      filename: IGVFFI3231EAMC.tsv.gz\n",
      "      filetype: tsv\n",
      "      filesize: 124\n",
      "      url: https://api.data.igvf.org/tabular-files/IGVFFI3231EAMC/@@download/IGVFFI3231EAMC.tsv.gz\n",
      "      urltype: https\n",
      "      md5: 91bb39e4be46f1ffb51caa4eb54f8897\n",
      "    regions: null\n",
      "    parent_id: tag\n",
      "  - !Region\n",
      "    region_id: MULTI-illumina_p7\n",
      "    region_type: illumina_p7\n",
      "    name: Illumina P7\n",
      "    sequence_type: fixed\n",
      "    sequence: ATCTCGTATGCCGTCTTCTGCTTG\n",
      "    min_len: 24\n",
      "    max_len: 24\n",
      "    onlist: null\n",
      "    regions: null\n",
      "    parent_id: tag\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(example_spec.to_YAML())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "centered-bookmark",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(run_print_library_tree(example_spec))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "parental-potato",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#_ = run_print_library_png(example_spec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef4bb624",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a31e71c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fa181521",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b7ef2f4d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_seqspecs_targz(table, destination):\n",
    "    fastqs = [\"R1\",\"R2\",\"R3\",\"I1\",\"I2\"]\n",
    "    \n",
    "    with tarfile.open(destination, \"w:gz\") as archive:\n",
    "        configuration_file = []\n",
    "        for i, row in tqdm.tqdm(table.iterrows(), total=table.shape[0]):\n",
    "            row_name = row[\"identifier\"].replace(\"ansuman-satpathy:\", \"\")\n",
    "            modality = row[\"modality\"]\n",
    "            #print(\"Processing {}\".format(row_name))\n",
    "            spec = create_seqspec(row)\n",
    "            spec_yaml = spec.to_YAML().encode(\"utf-8\")\n",
    "            spec_yaml = gzip.compress(spec_yaml, mtime=0)\n",
    "            spec_stream = BytesIO(spec_yaml)\n",
    "            spec_filename = f\"{row_name}_{modality}_seqspec.yaml.gz\"\n",
    "            spec_tarinfo = tarfile.TarInfo(str(spec_filename))\n",
    "            spec_tarinfo.size = len(spec_yaml)\n",
    "            archive.addfile(spec_tarinfo, fileobj=spec_stream)\n",
    "            configuration_file.append({\n",
    "                \"accession\": None,\n",
    "                \"file_set\": row[\"identifier\"],\n",
    "                \"content_type\": \"seqspec\",\n",
    "                \"file_format\": \"yaml\",\n",
    "                \"submitted_file_name\": spec_filename,\n",
    "                \"md5sum\": hashlib.md5(spec_yaml).hexdigest(),\n",
    "                \"seqspec_of:array\": \",\".join([row[x] for x in fastqs if pandas.notnull(row[x])]),\n",
    "                \"award\": award,\n",
    "                \"lab\": lab,\n",
    "                \n",
    "            })\n",
    "        configuration_file = pandas.DataFrame(configuration_file)\n",
    "        configuration_tsv = configuration_file.to_csv(index=False, sep=\"\\t\").encode(\"utf-8\")\n",
    "        configuration_stream = BytesIO(configuration_tsv)\n",
    "        configuration_info = tarfile.TarInfo(\"configuration_file.tsv\")\n",
    "        configuration_info.size = len(configuration_tsv)\n",
    "        archive.addfile(configuration_info, fileobj=configuration_stream)\n",
    "\n",
    "\n",
    "def save_team_seqspecs_directory(destination, table):\n",
    "    arguments = []\n",
    "    raw_url = \"https://raw.githubusercontent.com/detrout/y2ave_seqspecs/main/{filename}\"\n",
    "    \n",
    "    for i, row in table.iterrows():\n",
    "        row_name = row[0]\n",
    "        print(\"Processing {} {}\".format(team_id, row_name))\n",
    "        spec_filename = f\"{team_id}_{row_name}_seqspec.yaml\"\n",
    "        spec_yaml = render_seqspec(team_id, row)\n",
    "        with open(destination/spec_filename, \"wt\") as outstream:\n",
    "            outstream.write(spec_yaml)\n",
    "            \n",
    "        spec = load_spec_stream(StringIO(spec_yaml))\n",
    "        arg_row = {\n",
    "            \"team_id\": team_id,\n",
    "            \"dataset_id\": row_name,\n",
    "            \"url\": raw_url.format(filename=spec_filename),\n",
    "        }\n",
    "        print(generate_seqspec_index(spec, row))\n",
    "        arg_row.update(generate_seqspec_index(spec, row))\n",
    "        arguments.append(arg_row)\n",
    "\n",
    "    return arguments\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "a58883b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "#print(\"All\", metadata.shape[0])\n",
    "#metadata_rna_atac = metadata[metadata[\"modality\"].isin([\"atac\", \"rna\"])]\n",
    "#print(\"Removed multi modality\", metadata_rna_atac.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9441b16a",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#save_seqspecs_targz(metadata, \"mcginnis-seqspecs-20241125.tgz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67bffe68",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97bf5dfd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
